base/int.jl

```diff
@@ -282,63 +224,22 @@ const WORD_SIZE = Int(Int.size)*8
-
-promote_rule(::Type{UInt16},  ::Type{UInt8} ) = UInt16
-promote_rule(::Type{UInt32},  ::Type{UInt8} ) = UInt32
-promote_rule(::Type{UInt32},  ::Type{UInt16}) = UInt32
-promote_rule(::Type{UInt64},  ::Type{UInt8} ) = UInt64
-promote_rule(::Type{UInt64},  ::Type{UInt16}) = UInt64
-promote_rule(::Type{UInt64},  ::Type{UInt32}) = UInt64
-promote_rule(::Type{UInt128}, ::Type{UInt8} ) = UInt128
-promote_rule(::Type{UInt128}, ::Type{UInt16}) = UInt128
-promote_rule(::Type{UInt128}, ::Type{UInt32}) = UInt128
-promote_rule(::Type{UInt128}, ::Type{UInt64}) = UInt128
-
-promote_rule(::Type{UInt8}, ::Type{Int8}  ) = Int
-promote_rule(::Type{UInt8}, ::Type{Int16} ) = Int
-promote_rule(::Type{UInt8}, ::Type{Int32} ) = Int
-promote_rule(::Type{UInt8}, ::Type{Int64} ) = Int64
-promote_rule(::Type{UInt8}, ::Type{Int128}) = Int128
-
-promote_rule(::Type{UInt16}, ::Type{Int8}  ) = Int
-promote_rule(::Type{UInt16}, ::Type{Int16} ) = Int
-promote_rule(::Type{UInt16}, ::Type{Int32} ) = Int
-promote_rule(::Type{UInt16}, ::Type{Int64} ) = Int64
-promote_rule(::Type{UInt16}, ::Type{Int128}) = Int128
-
-if WORD_SIZE == 64
-    promote_rule(::Type{UInt32}, ::Type{Int8} ) = Int
-    promote_rule(::Type{UInt32}, ::Type{Int16}) = Int
-    promote_rule(::Type{UInt32}, ::Type{Int32}) = Int
-else
-    promote_rule(::Type{UInt32}, ::Type{Int8} ) = UInt
-    promote_rule(::Type{UInt32}, ::Type{Int16}) = UInt
-    promote_rule(::Type{UInt32}, ::Type{Int32}) = UInt
+promote_rule(::Type{Int8}, ::Type{Int16})   = Int16
+promote_rule(::Type{UInt8}, ::Type{UInt16}) = UInt16
+promote_rule{T<:Union{Int8,Int16}}(::Type{Int32}, ::Type{T})    = Int32
+promote_rule{T<:Union{UInt8,UInt16}}(::Type{UInt32}, ::Type{T}) = UInt32
+promote_rule{T<:Union{Int8,Int16,Int32}}(::Type{Int64}, ::Type{T})     = Int64
+promote_rule{T<:Union{UInt8,UInt16,UInt32}}(::Type{UInt64}, ::Type{T}) = UInt64
+promote_rule{T<:UnionS64Types}(::Type{Int128}, ::Type{T})    = Int128
+promote_rule{T<:UnionU64Types}(::Type{UInt128}, ::Type{T}) = UInt128
+for T in tuple(Signed64Types...,Int128)
+    @eval promote_rule{S<:Union{UInt8,UInt16}}(::Type{S}, ::Type{$T}) = $(sizeof(T) < sizeof(Int) ? Int : T)
 end
-promote_rule(::Type{UInt32}, ::Type{Int64} ) = Int64
-promote_rule(::Type{UInt32}, ::Type{Int128}) = Int128
-
-promote_rule(::Type{UInt64}, ::Type{Int8}  ) = UInt64
-promote_rule(::Type{UInt64}, ::Type{Int16} ) = UInt64
-promote_rule(::Type{UInt64}, ::Type{Int32} ) = UInt64
-promote_rule(::Type{UInt64}, ::Type{Int64} ) = UInt64
-promote_rule(::Type{UInt64}, ::Type{Int128}) = Int128
-
-promote_rule(::Type{UInt128}, ::Type{Int8}  ) = UInt128
-promote_rule(::Type{UInt128}, ::Type{Int16} ) = UInt128
-promote_rule(::Type{UInt128}, ::Type{Int32} ) = UInt128
-promote_rule(::Type{UInt128}, ::Type{Int64} ) = UInt128
-promote_rule(::Type{UInt128}, ::Type{Int128}) = UInt128
+@eval promote_rule{T<:Union{Int8,Int16,Int32}}(::Type{UInt32}, ::Type{T}) = $(WORD_SIZE == 64 ? Int : UInt)
```

Comment:
I'm not sure I understand - that one uses the eval to make whether it was `Int` or `UInt` at compile time instead of run-time, or will Julia take care of that?  That particular line was like that before, except in a loop, with `@eval`.